Add `nvm use` and auto-branching

message-dimke opened the pull request:
### Changes proposed in this Pull Request:

This PR extends `release` command with additional parameters
```
--nvm_use        - used to tell `release` to use `nvm use` per project basis.
--default_branch - used to pass project's default repo branch name into `release` command.
```
PR also adds few commands to be using as complementary commands to the overwritten `release` command:
```
simulate - used to run simulation mode on the new release command.
branch   - used to add a dialog for release branch creation.
```
Added set of tools for `Git`, `Nvm` and `Product`.

It is now possible to run the `release` and `simulate` commands using new parameters and see some changes in the command workflows.
```bash
./bin/wr simulate --nvm_use --default_branch develop --generate_changelog --wc_tested 8.3 --wp_tested 6.4 https://github.com/woocommerce/pinterest-for-woocommerce/tree/testrelease/1.3.13
```
For example if you provide a GitHub url using a branch name which does not exist on the repository, you will see a prompt to create it and release from there.

<img width="1551" alt="Terminal" src="https://github.com/woocommerce/grow/assets/9010963/1d1d5ee3-eb06-418d-8591-992a3ae66ef2">

It is also possible to skip release branch creation and release from the default which is set to `develop` by default. All you need is to refuse release branch creation.

<img width="1551" alt="Terminal" src="https://github.com/woocommerce/grow/assets/9010963/4cad2b22-6c56-463b-bb3f-9fea13ebe7ba">

The invisible part of changes introduced with this PR is `nvm use` command to run for every releasing product being inside the corresponding product directories to switch the release process to a proper `node` and `npm` versions. If `--nvm_use` flag is provided into `release` command, the additional logic will get involved into `build` process, if not - then the `release` will flow as usual.

### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Checkout `add-nvm-and-autobranching` branch.
2. Open `release.txt` in the root and add some lines to it.
```txt
pinterest-for-woocommerce	1.3.13	testrelease/1.3.13
woocommerce-brands	1.6.62	testrelease/1.6.62
```
**Note:** I am using Pinterest and Brands here because both projects have to use their own `node` version and w/o `nvm use` the build will fail. Pinterest can build with `node 14` when Brands require `node 16` for release.

You can provide any nonsense instead of a release branch names to check how branch prompt is working.
3. Run bulk command
```bash
./bin/wr bulk simulate --generate_changelog --wc_tested 8.3 --wp_tested 6.4
```
**Note:** I am using `bulk` command here as a single command to work with `release.txt`.  Also you may notice `--wc_tested` and `--wp_tested` flags, those are also here to check version bump still functional.
4. While releasing wait for a branch creation prompt to appear.
5. Choosing `y` will start a branch creation process. Branch will be created from a default one and pushed to remote. The push is requires for a `changelog` command to generate its changelog part.
6. Choosing `N` will drop release branch creation process and prompt to release from a default branch which is, if rejected, will terminate the release process.
7. When running the simulation every release branch, except the default, must be cleaned up automatically at the end of the process.
8. All the steps are followed with the corresponding console messages/notifications/errors.
### Additional details:

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

For changes related to the `github-actions` package, please use `[actions] changelog: *` labels.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

> Dev - Adding auto-release branching.
> Dev - Adding `nvm use` on product basis.

message-dimke committed:
- Extending `release` command.
- Adding default branches for extensions maintainer by Ventures.
- Replacing WooRelease `release` command with the local one.

message-dimke committed:
- Adding Git and Nvm command utility classes.
- Adjusting bulk command to call release with additional parameters.
- Adjusting release command to create release branch or fallback to the default repo branch and run nvm use if selected.

message-dimke committed:
Removing unused `use` statement.

message-dimke committed:
Removing test data.

JPry commented on extensions.json:
```diff
@@ -14,19 +14,22 @@
 			"name": "Points & Rewards",
-			"repoSlug": "woocommerce-points-and-rewards"
+			"repoSlug": "woocommerce-points-and-rewards",
+			"defaultBranch": "develop"
```
Since `develop` is now the default branch for our repos, we should instead change the `defaultBranch` property at the top of the file, and then we can skip all of these definitions.

JPry commented on packages/php/woorelease-extension/src/Utils/Nvm.php:
```diff
@@ -0,0 +1,47 @@
+	 */
+	public static function is_nvm_exists() {
+		try {
+			Utils::exec_sprintf( 'nvm -v' );
```
```suggestion
			Utils::exec_sprintf( 'which -s nvm' );
```

message-dimke replied:
Hey, @JPry ! I found `-s` is not an option on ubuntu for example. There is some inconsistency between operational systems. Do we still want it?
https://linux.die.net/man/1/which
https://ss64.com/osx/which.html

message-dimke replied:
Also, it is interesting but `which -s nvm` was working for me but then stopped. Running the command now gives me a function nvm() {} output into my console. I googled this:
```
The nvm command is not a file but rather a shell function.
The source ~/.nvm/nvm.sh adds these functions to your current shell. And because these commands are not files, they don't show up when you which nvm.
Looking in the .nvm/nvm.sh file, you can see a nvm() {...} function defined that provides that functionality.
```

message-dimke replied:
Also in Terminal I have zsh and exec uses sh. That is the reason I see some inconsistencies in commands when running from both places.

message-dimke replied:
Closing the conversation because of another way of solving the issue introduced with `/bin/build` and `/bin/nvm` shell scripts.

JPry commented on packages/php/woorelease-extension/src/Commands/Release.php:
```diff
@@ -0,0 +1,277 @@
+	public function __construct( string $name = null ) {
+		parent::__construct( $name );
+	}
+
```
We can omit all of this since we're not actually making changes.

```suggestion
```

JPry commented on .gitignore:
```diff
@@ -8,4 +8,5 @@ vendor/
-.vscode/*
\ No newline at end of file
+.vscode/*
+.idea
```
This should be part of your global `.gitignore`. 

```suggestion
```

JPry commented on bin/wr:
```diff
@@ -3,6 +3,7 @@
 
 use Automattic\WooCommerce\Grow\WR\Commands\Bulk;
 use Automattic\WooCommerce\Grow\WR\Commands\VersionReplace;
+use Automattic\WooCommerce\Grow\WR\Commands\Release;
```
Minor: it's nice to keep them in alphabetical order 🙂 

```suggestion
use Automattic\WooCommerce\Grow\WR\Commands\Release;
use Automattic\WooCommerce\Grow\WR\Commands\VersionReplace;
```

JPry commented on packages/php/woorelease-extension/src/Utils/Git.php:
```diff
@@ -0,0 +1,46 @@
+	 *
+	 * @return bool True if the branch exists, false otherwise.
+	 */
+	public static function is_branch_exists( $repository_url, $branch ) {
```
Changing the name helps this to read better.

```suggestion
	public static function does_branch_exist( $repository_url, $branch ) {
```

JPry commented on packages/php/woorelease-extension/src/Utils/Git.php:
```diff
@@ -0,0 +1,46 @@
+		}
+	}
+
+	public static function create_branch( $repository_url, $branch ) {
```
The `$repository_url` parameter isn't used, and we're missing a docblock.

JPry commented on packages/php/woorelease-extension/src/Utils/Nvm.php:
```diff
@@ -0,0 +1,47 @@
+	 *
+	 * @return bool True if `nvm` is installed, false otherwise.
+	 */
+	public static function is_nvm_exists() {
```
Changing the name helps this to read better.

```suggestion
	public static function does_nvm_exist() {
```

JPry reviewed (commented):
I know this is progress, and I'm not able to dedicate time to a full review right now, but I just wanted to leave a few comments.

message-dimke committed:
Overwriting the simulate command.

message-dimke committed:
- Adding branch command.
- Renaming some method names to be grammatically correct.
- Separating create and push branch commands into different methods.

message-dimke committed:
Removing unnecessary props and methods.

message-dimke committed:
Changing default branch name for all the extensions since that is de facto a standard  now.

message-dimke committed:
Removing IDE unique files and folders from .gitignore

message-dimke committed:
Fixing wrong placed commas after removing default branch names from extensions.

message-dimke committed:
Remove redundant release command option. When we are releasing we are providing a release branch name which, if does not exist, is prompted to get created or fallbacks to default. This means the parameter to create a release branch is redundant.

message-dimke committed:
Adding branch command method commends and removing redundant constructor from the command class.

message-dimke committed:
Adjusting `branch` command code styling and output.

message-dimke committed:
Optimising imports (doing them in alphabetical order)

message-dimke committed:
Removing Simulate command since we have nothing to overwrite with it.

message-dimke committed:
Adding `nvm` shell script to fight some OS shell inconsistencies. Adjust commands to use the newly added shell script. Adding custom `simulate` command which inherits from the overwritten `release` command to allow simulation runs. Update `branch` command by adding cleanup option to it.

message-dimke committed:
Adding build script to run nvm and npm build steps in a single context. Create custom product build function to handle cases with and w/o nvm.

message-dimke committed:
Adding build script to set nvm and npm into a single context when running the build. Using application's meta storage to pass branches between `branch` and `release` commands. Adjusting release branches cleanup to make sure not to cleanup the default branch from the product's repo.

rawdreeg reviewed (approved):
Sorry for the delay. This has been in my to-do list for a while lol.

Looks good, all the previous feedback and nits are now fixed, and I couldn't find any other issue.

While testing this I encountered #99 and I submitted a PR fixing it (#100). So I'll approve this!

message-dimke committed:
- Fixing branch operations;
- Updating `woorelease` dev-trunk dependency version since the last release was back 2022.

message-dimke committed:
Merge remote-tracking branch 'refs/remotes/origin/trunk' into add-nvm-and-autobranching

# Conflicts:
#	composer.json
#	composer.lock
#	packages/php/woorelease-extension/src/Commands/Bulk.php

message-dimke committed:
Adding recent composer lock file.

message-dimke committed:
Adding composer lock file ran with php 7.4

message-dimke committed:
Fixing phpcs issues

message-dimke committed:
Fixing phpcs issues with spaces and @throws missing.

message-dimke committed:
Fixing phpcs issues.

message-dimke committed:
Adding node auto-install command.
